In [2]:
import pandas as pd
import numpy as np


In [3]:
import os

os.chdir('..')
os.getcwd()

'c:\\Users\\k\\Problem-Solving\\Topics\\Statistics\\3_Regression\\Linear_regression'

In [4]:
df = pd.read_csv('data.csv')
df.head()

,area_m2,rooms,bedrooms,bathrooms,floor,total_floors,building_age_years,distance_city_center_km,distance_public_transport_km,school_rating,crime_index,parks_within_1km,parking_spaces,renovation_score,energy_efficiency_score,noise_level,neighborhood_income_index,has_elevator,is_new_building,apartment_price_usd
0,64.6,3,2,1,9,25,11.7,3.65,1.18,8.1,29.6,0,1,8.5,8.7,43.2,57.4,1,0,402033.11
1,136.7,6,5,2,3,24,26.5,4.13,0.46,8.7,8.2,3,1,7.5,4.1,44.8,74.5,1,0,716709.88
2,99.5,4,3,1,10,29,8.4,5.69,0.71,6.9,30.2,4,1,7.5,8.6,40.0,73.8,0,0,569233.62
3,86.2,4,3,2,25,30,23.8,13.50,0.44,7.1,40.3,0,1,7.6,6.7,41.2,42.7,1,0,388190.73
4,73.7,2,1,2,1,5,63.1,6.55,2.18,6.9,45.6,1,0,4.6,7.0,46.8,65.4,0,0,296126.18


In [5]:
X = df.drop(columns=['apartment_price_usd'])
y = df['apartment_price_usd']

n_samples = X.shape[0]

np.random.seed(42)
indices = np.random.permutation(n_samples)

X_shuffled = X.iloc[indices]
y_shuffled = y.iloc[indices]

X_train = X_shuffled[:4000]
y_train = y_shuffled[:4000]

X_test = X_shuffled[4000:]
y_test = y_shuffled[4000:]

print(f"Training Data: {X_train.shape}")
print(f"Training Data Target: {y_train.shape}")
print(f"Test Data: {X_test.shape}")
print(f"Test Data Target: {y_test.shape}")
n_samples

Training Data: (4000, 19)
Training Data Target: (4000,)
Test Data: (1000, 19)
Test Data Target: (1000,)


5000

In [6]:
def standardization(x):
    mean = x.mean()
    std = x.std()

    x_stand = (x - mean) / std

    return x_stand

X_train_stand = standardization(x=X_train)
X_test_stand = standardization(x=X_test)

In [7]:
ones = np.ones((X_train.shape[0], 1))
X_with_intercept = np.hstack((ones, X_train))

beta_hat = np.linalg.pinv(X_with_intercept.T @ X_with_intercept ) @ X_with_intercept.T @ y_train

intercept = beta_hat[0]
coefficients = beta_hat[1:]

In [8]:
parameters = pd.DataFrame({
    "Feature": ["Intercept"],
    "Value": [intercept]
})

for feature, coefficient in zip(X.columns, coefficients):
    parameters.loc[len(parameters)] = [feature, coefficient]

parameters

,Feature,Value
0,Intercept,38013.120221
1,area_m2,2831.815021
2,rooms,1699.504703
3,bedrooms,8563.932251
4,bathrooms,14508.320363
5,floor,893.422817
6,total_floors,254.235664
7,building_age_years,-1564.070681
8,distance_city_center_km,-6228.041664
9,distance_public_transport_km,-8733.583433


In [9]:
train_predictions = X_train @ coefficients + intercept
test_predictions = X_test @ coefficients + intercept

# Interpreting Coefficients

## Assessing the significance of coefficients

## Evaluating magnitude and direction of coefficients

In [10]:
parameters

,Feature,Value
0,Intercept,38013.120221
1,area_m2,2831.815021
2,rooms,1699.504703
3,bedrooms,8563.932251
4,bathrooms,14508.320363
5,floor,893.422817
6,total_floors,254.235664
7,building_age_years,-1564.070681
8,distance_city_center_km,-6228.041664
9,distance_public_transport_km,-8733.583433


- **Intercept** - `38013.12`: predicted apartment price when all numerical features are 0 and binary features are 0; mainly mathematical meaning.
- **area_m2** - `2831.82`: Magnitude = $2,831.82 per 1 m², Direction = Positive; +1 m² increases predicted apartment price by about $2,832.
- **rooms** - `1699.50`: Magnitude = $1,699.50, Direction = Positive;  +1 room increases predicted apartment price by about $1,700.
- **bedrooms** - `8563.93`: Magnitude = $8,563.93, Direction = Positive; +1 bedroom increases predicted apartment price by about $8,564.
- **bathrooms** - `14508.32`: Magnitude = $1,4508.32, Direction = Positive; +1 bathroom increases predicted apartment price by about $14,508.
- **floor** - `893.42`:Magnitude = $893.42, Direction = Positive; +1 floor level increases predicted apartment price by about $893.
- **total_floors** - `254.24`: Magnitude = $254.24, Direction = Positive; +1 total floor in the building increases predicted apartment price by about $254.
- **building_age_years** - `-1564.07`: Magnitude = $-1,564.07, Direction = Negative; +1 year of building age decreases predicted apartment price by about $1,564.
- **distance_city_center_km** - `-6228.04`: Magnitude = $-6228.04, Direction = Negative; +1 km farther from the city center decreases predicted apartment price by about $6,228.
- **distance_public_transport_km** - `-8733.58`: Magnitude = $-8733.58, Direction = Negative; +1 km farther from public transport decreases predicted apartment price by about $8,734.
- **school_rating** - `7154.05`: Magnitude = $7154.05, Direction = Positive; +1 point in school rating increases predicted apartment price by about $7,154.
- **crime_index** - `-926.56`: Magnitude = -926.56, Direction = Negative; +1 point in crime index decreases predicted apartment price by about $927.
- **parks_within_1km** - `4219.59`: Magnitude = $4219.59, Direction = Positive; +1 additional park within 1 km increases predicted apartment price by about $4,220.
- **parking_spaces** - `12304.13`: Magnitude = $12304.13, Direction = Positive; +1 parking space increases predicted apartment price by about $12,304.
- **renovation_score** - `4901.83`: Magnitude = $4901.83, Direction = Positive; +1 point in renovation score increases predicted apartment price by about $4,902.
- **energy_efficiency_score** - `3487.02`: Magnitude = $3487.02, Direction = Positive; +1 point in energy-efficiency score increases predicted apartment price by about $3,487.
- **noise_level** - `-896.09`: Magnitude = $-896.09, Direction = Negative; +1 point in noise level decreases predicted apartment price by about $896.
- **neighborhood_income_index** - `2090.28`: Magnitude = $2090.28, Direction = Positive; +1 point in neighborhood-income index increases predicted apartment price by about $2,090.
- **has_elevator** - `18020.90`: Magnitude = $18020.90, Direction = Positive; if 0 = No and 1 = Yes, having an elevator increases predicted apartment price by about $18,021.
- **is_new_building** - `21397.84`: Magnitude = $21397, Direction = Positive; if 0 = No and 1 = Yes, being a new building increases predicted apartment price by about $21,398.

# Assessing the model fit

## R2-square

In [11]:
def R2_score(y_true, y_pred):
    if len(y_true) != len(y_pred):
            raise ValueError("The length of y_true and y_pred must be the same.")
    SSR = np.sum((y_true - y_pred)**2)
    SST = np.sum((y_true - np.mean(y_true))** 2)
    return 1 - (SSR / SST)

In [12]:
r2_score_train = R2_score(y_train, train_predictions)
r2_score_train

np.float64(0.9603881139719969)

In [13]:
r2_score_test = R2_score(y_test, test_predictions)
r2_score_test

np.float64(0.96074175355072)

In [14]:
def adjusted_R2_score(y_true, y_pred, X_train):
    if len(y_true) != len(y_pred):
            raise ValueError("The length of y_true and y_pred must be the same.")

    n, p = X_train.shape
    SSR = np.sum((y_true - y_pred) ** 2)
    SST = np.sum((y_true - np.mean(y_true)) ** 2)
    ordinary_R2 = 1 - (SSR / SST)
    return 1 - (1 - ordinary_R2) * ((n - 1) / (n - p - 1))

In [15]:
adj_r2_score_train = adjusted_R2_score(y_train, train_predictions, X_train)
adj_r2_score_train

np.float64(0.9601990120035214)

In [16]:
adj_r2_score_test = adjusted_R2_score(y_test, test_predictions, X_test)
adj_r2_score_test

np.float64(0.9599806242828258)

## Root mean squared error RMSE

In [17]:
def root_mean_squared_error(y_true, y_pred):
    if len(y_true) != len(y_pred):
                raise ValueError("The length of y_true and y_pred must be the same.")
    
    n = len(y_pred)
    rmse = np.sqrt(np.sum((y_true - y_pred)**2) / n)
    return rmse

In [18]:
rmse_train = root_mean_squared_error(y_train, train_predictions)
rmse_train

np.float64(21691.301585996516)

In [19]:
rmse_test = root_mean_squared_error(y_test, test_predictions)
rmse_test

np.float64(21714.99074042335)

## F-statistic

In [20]:
def f_statistic(y_true, y_pred, X_train):
    SSR = np.sum((np.mean(y_true) - y_pred) ** 2)
    SSE = np.sum((y_pred - y_true) ** 2)

    n, p = X_train.shape
    regression_df  = p
    error_of_freedom = n - p - 1

    MSR = SSR / regression_df 
    MSE = SSE / error_of_freedom

    f_statistic = MSR / MSE 
    return f_statistic

In [21]:
f_statistic_train = f_statistic(y_train, train_predictions, X_train)
f_statistic_train

np.float64(5078.6785654075575)

In [22]:
f_statistic_test = f_statistic(y_test, test_predictions, X_test)
f_statistic_test

np.float64(1277.9089157236854)

## AIC

In [23]:
def aic_calculation(y_true, y_pred, X):
    n = len(y_true)
    p = X.shape[1]

    SSE = np.sum((y_true - y_pred) ** 2)

    sigma_squared = SSE / n

    log_likelihood = -(n / 2) * (
        np.log(2 * np.pi)
        + np.log(sigma_squared)
        + 1 
    )

    k = p + 1

    AIC = 2 * k - 2 * log_likelihood
    return AIC

### Model-1 

features: `area_m2`

In [24]:
X_area = df[['area_m2']]
y_target = df['apartment_price_usd']

In [25]:
ones = np.ones((X_area.shape[0], 1))
X_area_with_intercept = np.hstack((ones, X_area))

beta_area = np.linalg.pinv(X_area_with_intercept.T @ X_area_with_intercept) @ X_area_with_intercept.T @ y_target
intercept_area = beta_area[0]
coefficient_area = beta_area[1:]

In [26]:
area_model_predictions = X_area @ coefficient_area + intercept_area


In [27]:
model_1_AIC = aic_calculation(y_target, area_model_predictions, X_area)

In [28]:
aic_results = pd.DataFrame({
    "model": "Model-1",
    "AIC": model_1_AIC,
    "features":['area_m2']
}, index=[0])
aic_results

,model,AIC,features
0,Model-1,125706.056474,area_m2


### Model-2

features: apartment size and structure 

In [29]:
model_2_features = [
    "area_m2",
    "rooms",
    "bedrooms",
    "bathrooms"
]

X_model_2 = df[model_2_features]

In [30]:
X_model_2_with_intercept = np.hstack((ones, X_model_2))
beta_model_2 = np.linalg.pinv(X_model_2_with_intercept.T @ X_model_2_with_intercept) @ X_model_2_with_intercept.T @ y_target
intercept_model_2 = beta_model_2[0]
coefficients_model_2 = beta_model_2[1:]

In [31]:
predictions_model_2 = X_model_2 @ coefficients_model_2 + intercept_model_2

In [32]:
aic_model_2 = aic_calculation(y_target, predictions_model_2, X_model_2)
aic_results.loc[len(aic_results)] = ['model-2',aic_model_2, model_2_features]

In [33]:
aic_results

,model,AIC,features
0,Model-1,125706.056474,area_m2
1,model-2,125671.881116,"[area_m2, rooms, bedrooms, bathrooms]"


### Model-3

features: apartment size + structure + characteristics

In [34]:
model_3_features = [
    "area_m2",
    "rooms",
    "bedrooms",
    "bathrooms",
    "floor",
    "total_floors",
    "building_age_years",
    "has_elevator",
    "is_new_building"
]


X_model_3 = df[model_3_features]

In [35]:
X_model_3_with_intercept = np.hstack((ones, X_model_3))

beta_model_3 = np.linalg.pinv(
    X_model_3_with_intercept.T @ X_model_3_with_intercept
) @ X_model_3_with_intercept.T @ y_target

intercept_model_3 = beta_model_3[0]
coefficients_model_3 = beta_model_3[1:]

In [36]:
predictions_model_3 = X_model_3 @ coefficients_model_3 + intercept_model_3

In [37]:
aic_model_3 = aic_calculation(y_target, predictions_model_3, X_model_3)
aic_results.loc[len(aic_results)] = ['Model-3', aic_model_3, model_3_features]
aic_results

,model,AIC,features
0,Model-1,125706.056474,area_m2
1,model-2,125671.881116,"[area_m2, rooms, bedrooms, bathrooms]"
2,Model-3,124007.849984,"[area_m2, rooms, bedrooms, bathrooms, floor, t..."


### Model-4

features: area + strucutre + characteristics + location

In [38]:
model_4_features = [
    "area_m2",
    "rooms",
    "bedrooms",
    "bathrooms",
    "floor",
    "total_floors",
    "building_age_years",
    "has_elevator",
    "is_new_building",
    "distance_city_center_km",
    "distance_public_transport_km",
    "school_rating",
    "crime_index",
    "parks_within_1km",
    "noise_level",
    "neighborhood_income_index"
]

In [39]:
X_model_4 = df[model_4_features]


In [40]:
X_model_4_with_intercept = np.hstack((ones, X_model_4))
beta_model_4 = np.linalg.pinv(
    X_model_4_with_intercept.T @ X_model_4_with_intercept
) @ X_model_4_with_intercept.T @ y_target

intercept_model_4 = beta_model_4[0]
coefficients_model_4 = beta_model_4[1:]

In [41]:
predictions_model_4 = X_model_4 @ coefficients_model_4 + intercept_model_4 

In [42]:
aic_model_4 = aic_calculation(y_target, predictions_model_4, X_model_4)
aic_results.loc[len(aic_results)] = ["Model-4", aic_model_4, model_4_features]
aic_results

,model,AIC,features
0,Model-1,125706.056474,area_m2
1,model-2,125671.881116,"[area_m2, rooms, bedrooms, bathrooms]"
2,Model-3,124007.849984,"[area_m2, rooms, bedrooms, bathrooms, floor, t..."
3,Model-4,115172.960992,"[area_m2, rooms, bedrooms, bathrooms, floor, t..."


## BIC 

In [43]:
def bic_calculation(y_true, y_pred, X):
    n = len(y_true)
    p = X.shape[1]

    SSE = np.sum((y_true - y_pred) ** 2)

    sigma_squared = SSE / n

    log_likelihood = -(n / 2) * (
        np.log(2 * np.pi)
        + np.log(sigma_squared)
        + 1
    )
    k = p + 1

    BIC = k * np.log(n) - 2 * log_likelihood
    return BIC


# Bias-Variance Tradeoff

## Overfitting

In [51]:
results = pd.DataFrame({
    "Metric":"RMSE",
    "Value":rmse_train,
    "Data":"Train"
}, index=[0])
results

,Metric,Value,Data
0,RMSE,21691.301586,Train


In [52]:
results.loc[len(results)] = ["RMSE", rmse_test, 'Test']
results.loc[len(results)] = ["R2", r2_score_train, 'Train']
results.loc[len(results)] = ["R2", r2_score_test, 'Test']

In [54]:
results.loc[len(results)] = ["F-score", f_statistic_train , 'Train']
results.loc[len(results)] = ["F-score", f_statistic_test, 'Test']

In [55]:
results

,Metric,Value,Data
0,RMSE,21691.301586,Train
1,RMSE,21714.990740,Test
2,R2,0.960388,Train
3,R2,0.960742,Test
4,F-score,5078.678565,Train
5,F-score,1277.908916,Test
